src/aiokatcp/time_sync.py

```diff
@@ -0,0 +1,110 @@
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+# DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+"""Utilities for creating time synchronisation sensors.
+
+These use the Linux-specific :func:`adjtimex` system call to get information
+about time synchronisation.  This does not give access to all the statistics
+that an NTP or PTP server would provide, but should be sufficient to check that
+time is being kept synchronised, and (at least on Linux) is accessible from
+inside a container with no privileges required.
+"""
+
+import asyncio
+from enum import Enum
+from typing import Mapping
+
+from . import adjtimex
+from .sensor import Sensor
+
+
+class ClockState(Enum):
+    # Note: values must match TIME_* constants in adjtimex.py
+    OK = 0
+    INS = 1
+    DEL = 2
+    OOP = 3
+    WAIT = 4
+    ERROR = 5
+
+
+def _us_to_s(value):
+    return 1e-6 * value
+
+
+_TRANSFORMS = {
+    "maxerror": _us_to_s,
+    "esterror": _us_to_s,
+    "state": ClockState,
+}
+
+
+class TimeSyncUpdater:
+    """Maps raw adjtimex(2) fields to sensor values.
+
+    The sensors to populate are specified by a mapping, whose keys can be any
+    subset of:
+
+    state
+        Return value of the :c:func:`!adjtimex` call (as a :class:`ClockState`)
+    maxerror
+        Maximum error, in seconds
+    esterror
+        Estimated error, in seconds
+    """
+
+    def __init__(self, sensor_map: Mapping[str, Sensor]) -> None:
+        self.sensor_map = dict(sensor_map)
+        for key in self.sensor_map.keys():
+            if key not in _TRANSFORMS:
+                raise KeyError(f"Key {key!r} is not valid")
+
+    def update(self) -> None:
+        """Update the sensors now."""
+        state, timex = adjtimex.get_adjtimex()
+        if timex.status & adjtimex.STA_NANO:
+            scale = 1e-9
+        else:
+            scale = 1e-6
+        now = timex.time.tv_sec + scale * timex.time.tv_usec
+        for key, sensor in self.sensor_map.items():
+            if key == "state":
+                value = state
+            else:
+                value = getattr(timex, key)
+            value = _TRANSFORMS[key](value)
+            sensor.set_value(value, timestamp=now)
+
+    async def run(self, period: float) -> None:
+        """Run until cancelled, updating the sensors every `period` seconds.
+
+        In the current implementation the `period` won't be exact, because it
+        doesn't account for the time spent performing the update.
```

Comment:
The update is synchronous though, so I don't think that works.

It's definitely solveable, but
(a) I don't expect the update part to take more than a millisecond, whereas I'm thinking that a reasonable update period is 1-5 seconds.
(b) It really doesn't matter if there is a bit of drift, and it may even be preferable to avoid a thundering herd if a bunch of engines start up in sync.

I'm also thinking it might be worth deleting this function for now. I'm not actually using it in katgpucbf, because there I want to customise the update loop to also compute a composite `time.synchronised` flag at the same time.

Reply:
> I'm also thinking it might be worth deleting this function for now. I'm not actually using it in katgpucbf [...]

I `gather` (ha!) the katgpucbf branch you mentioned in the check-in addresses this more custom, composite `time.synchronised`. When you say 'delete this (`run`) function', do you mean 
* With expectation for it to be added back in?
* Or, with expectation that it's up to the user to implement this (maybe abstract?) method?

EDIT: Also, @james-smith-za if you had any thoughts on Bruce's suggestion.

Reply:
With the expectation that if someone asks for it, we can add it later, while not adding things until they're needed.

Reply:
Yeah I'm absolutely for this. This == the removal of the function until someone/some use case requires it.

Reply:
If it's not getting used then it's probably best to remove it for now. It can be added back later if we need it.

Reply:
Done.